core/state/statedb.go

```diff
@@ -760,6 +761,11 @@ func (db *DB) Commit(deleteEmptyObjects bool) (root common.Hash, err error) {
 		if obj := db.stateObjects[addr]; !obj.deleted {
 			// Write any contract code associated with the state object
 			if obj.code != nil && obj.dirtyCode {
+				if !commitValidatorWrapper && obj.IsValidator(db.db) {
+					db.deleteStateObject(obj)
+					obj.deleted = true
```

Comment:
Unfortunately, from my knowledge, if you delete the state object, all information of the address will be deleted, which will make the address unrecoverable. Have you tested the case whether the address is still accessible after this operation?

Also, since the commit block for archive node is not in consensus, could it happen that different archive nodes will have different validatorWrapper commit block, thus resulting in different block hash?

Reply:
Yes. the address still accessible after this operation, because latest state will keep in memory. So for the Flush to disk operation, I used another copy of the state to modify and commit, but the latest state trie was kept in memory. But for history blocks, you can't rely on the state trie in memory, you need to see if the block is committed to disk. Otherwise, it will be empty.  you can see my test. Definitely does not affect the block hash, because this root value is calculated and only used to commit the state, and does not modify the root in the block header.    Please wait , I will check agian!

Reply:
From my knowledge, the state to be stored actually is a different state with a different root from the state in the block. I doubt whether any part of the state from block is recoverable after committing a different state like this. Would you please write a unit test to address my concern?

Reply:
Yes, I know that the state to be stored actually is a different state with a different root from the state in the block, but the goal is that do not save validatorwrapper, it must be different root state from state in the block. if keep root state consistent with block, the validatorwrapper will commited to disk for every block

Reply:
Yes I will write unit test to check

Reply:
My understanding is, if the states of 127/128 blocks in archival node are not in storage, thus not retrievable, what is the point of starting an archival node in the first place?

Reply:
you are right! The goal of this issue is essentially a compromise on the archive node state explosion, a workaround, to do perfection is unlikely. A large number of Tx state changes are not discardable. As discussed with RJ before, the best way is to store the members of ValidatorWrapper separately, but that would be complicated. So I have discussed with RJ, and the simple idea is to swipe the disk once every N blocks, but then the state of the Validator cannot be checked on some blocks, my concern is: I don't know if the test can cover all, because the root state is really inconsistent, there may be a corner case that I have not considered
 
do you have any better approaches  to handle  state explosion on archive node?